drafts/two-way-binding-signals-and-query-params.md

```diff
@@ -0,0 +1,754 @@
+```ts
+@Injectable({ providedIn: "root" })
+export class QueryParamService {
+  private activatedRoute = inject(ActivatedRoute);
+
+  get(queryParamName: string) {
+    const value$ = this.activatedRoute.queryParams.pipe(
+      map((allQueryParams) => allQueryParams[queryParamName])
+    );
+
+    return toSignal(value$);
+  }
+}
+```
+
+```ts
+@Component({
+  selector: "app-root",
+  standalone: true,
+  template: `
+    <h1>Signals Demo</h1>
+    <p>Full name: {{ fullName() }}</p>
+    <p>Age: {{ age() }}</p>
+
+    <button (click)="anonymise()">I want to be anonymous 👤</button>
+  `,
+})
+export class AppComponent {
+  private queryParamService = inject(QueryParamService);
+
+  firstName = this.queryParamService.get("firstName");
+  lastName = this.queryParamService.get("lastName");
+  age = this.queryParamService.get("age");
+
+  fullName = computed(() => `${this.firstName() ?? ""} ${this.lastName() ?? ""}`);
+
+  anonymise() {
+    // TODO: in a minute :)
+  }
+}
+```
+
+This small refactor will produce the same behavior as before except the anonymise feature, which we will implement in the nicer way as soon as we make our `get` method from `QueryParamService` return a writable signal.
+
+This is a very interesting challenge. So let's start simple:
+
+```ts
+@Injectable({ providedIn: "root" })
+export class QueryParamService {
+  private activatedRoute = inject(ActivatedRoute);
+
+  get(queryParamName: string) {
+    const queryParamValue = signal("");
+
+    // TODO: connect signal with actual query param value.
+
+    return queryParamValue;
+  }
+}
+```
+
+We need to return a writable signal, let's call it `queryParamValue`. Right now it is completely disconnected from the corresponding query param. That's the problem we want to solve.
+
+More specifically, we need to connect it to the `activatedRoute.queryParams` observable, but manually creating and maintaining a subscription in our service doesn't sound like a signal-friendly idea:
+
+```ts
+@Injectable({ providedIn: "root" })
+export class QueryParamService {
+  private activatedRoute = inject(ActivatedRoute);
+
+  get(queryParamName: string) {
+    const queryParamValue = signal("");
+
+    // This will "work" but we don't want to explicitly subscribe here ❌
+    this.activatedRoute.queryParams.subscribe((allQueryParams) =>
+      queryParamValue.set(allQueryParams[queryParamName])
+    );
+
+    return queryParamValue;
+  }
+}
+```
+
+A signal-friendly idea could be to convert the `activatedRoute.queryParams` observable to a signal itself and then somehow synchronize it with `queryParamValue`:
+
+```ts
+@Injectable({ providedIn: "root" })
+export class QueryParamService {
+  private activatedRoute = inject(ActivatedRoute);
+  private allQueryParams = toSignal(this.activatedRoute.queryParams);
+
+  get(queryParamName: string) {
+    const queryParamValue = signal("");
+
+    // TODO: sync queryParamValue signal with allQueryParams signal
+
+    return queryParamValue;
+  }
+}
+```
```

Comment:
**suggestion:** skip this code, it doesn't add much here, just have the two paragraphs and then code. For an article, too many code snippets will make it longer and harder to read. Readers will eventually skip, if they can't pinpoint the exact changes quickly.